### PR TITLE
Fix write_json failure for zero columns in table/struct

### LINKED ISSUE #17413: [BUG] write_json fails for struct with zero columns
**Describe the bug**
When a table has struct with zero columns, it fails to write using `cudf::write_json`.

**Steps/Code to reproduce bug**
case 1:
```python
# write a json with struct with zero columns, sometimes nulls
json_str = """
{"A": {"B": null, "C": 1}}
{"A": {"B": null}}
{"A": {"B": {}}}
{"A": {"B": {}}}
{"A": {}}
{"A": {}}
"""
import cudf
from io import StringIO
df1 = cudf.read_json(StringIO(json_str), lines=True)
df1[["A"]].to_json('tmp', engine='cudf')
```

Also, another repro (different issue).
case 2:
```python
# write a json with struct with zero columns, sometimes nulls
json_str = """
{"A": null}
{"A": null}
{"A": {}}
{"A": {}}
{}
{}
"""
import cudf
from io import StringIO
df1 = cudf.read_json(StringIO(json_str), lines=True)
df1[["A"]].to_json('tmp', engine='cudf')
```

**Expected behavior**
using pandas engine, `df1[["A"]].to_json()` produces

case 1: `'{"A":{"0":{"B":null,"C":1},"1":{"B":null,"C":null},"2":{"B":{},"C":null},"3":{"B":{},"C":null},"4":{"B":null,"C":null},"5":{"B":null,"C":null}}}'`
case 2: `'{"A":{"0":null,"1":null,"2":{},"3":{},"4":null,"5":null}}'`

**Additional context**
@GregoryKimball found the issue while working with synthesized json files.

### PR DESCRIPTION
**Pull request by karthikeyann**
## Description
Closes  #17413

num_rows are passed to ensure empty`{}` is created for zero columns.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/src/io/json/write_json.cu [hunk 1]
```diff
@@ -268,40 +270,53 @@ std::unique_ptr<column> struct_to_strings(table_view const& strings_columns,
...
+  if (num_rows == 0)  // empty begets empty
     return make_empty_column(type_id::STRING);
   // check all columns are of type string
   CUDF_EXPECTS(std::all_of(strings_columns.begin(),
                            strings_columns.end(),
                            [](auto const& c) { return c.type().id() == type_id::STRING; }),
                "All columns must be of type string");
   auto constexpr strviews_per_column = 3;  // (for each "column_name:", "value",  "separator")
-  auto const num_strviews_per_row    = strings_columns.num_columns() * strviews_per_column + 1;
+  auto const num_strviews_per_row    = 1 + strings_columns.num_columns() * strviews_per_column -
```
Comment (bdice) [inline cpp/src/io/json/write_json.cu:281]: Can we avoid the subtraction here? It looks like we have “maybe subtract 1, then add 1” and that might be better to write as “maybe add 1” with a reversed condition.

Comment (karthikeyann) [reply cpp/src/io/json/write_json.cu:281]: updated.

File: cpp/src/io/json/write_json.cu [hunk 2]
```diff
@@ -268,40 +270,53 @@ std::unique_ptr<column> struct_to_strings(table_view const& strings_columns,
...
+                                         include_nulls,
+                                         d_strviews.begin()};
+    // scatter row_prefix, row_suffix, column_name:, value, value_separator as string_views
+    thrust::for_each(rmm::exec_policy(stream),
+                     thrust::make_counting_iterator<size_type>(0),
+                     thrust::make_counting_iterator<size_type>(total_rows),
+                     scatter_fn);
+  } else {
+    thrust::for_each(
+      rmm::exec_policy(stream),
```
Comment (vuule) [inline cpp/src/io/json/write_json.cu:310]: can we do nosync here? (maybe in the other branch as well)
```suggestion
      rmm::exec_policy_nosync(stream),
```

File: cpp/src/io/json/write_json.cu [hunk 3]
```diff
@@ -268,40 +270,53 @@ std::unique_ptr<column> struct_to_strings(table_view const& strings_columns,
...
     return make_empty_column(type_id::STRING);
   // check all columns are of type string
   CUDF_EXPECTS(std::all_of(strings_columns.begin(),
                            strings_columns.end(),
                            [](auto const& c) { return c.type().id() == type_id::STRING; }),
                "All columns must be of type string");
   auto constexpr strviews_per_column = 3;  // (for each "column_name:", "value",  "separator")
-  auto const num_strviews_per_row    = strings_columns.num_columns() * strviews_per_column + 1;
+  auto const num_strviews_per_row    = 1 + strings_columns.num_columns() * strviews_per_column +
+                                    (strings_columns.num_columns() == 0 ? 1 : 0);
```
Comment (vuule) [inline cpp/src/io/json/write_json.cu:282]: I can't tell what the first 1 is, and what the 1 added for `strings_columns.num_columns() == 0` is

Comment (karthikeyann) [reply cpp/src/io/json/write_json.cu:282]: How about:
`strings_columns.num_columns() == 0 ? 2 : 1 + strings_columns.num_columns() * strviews_per_column;`?
2 for empty `{}`

Which is essentially `1 + strings_columns.num_columns() * strviews_per_column - 1 + 1`;
for ` {col1: value, col2: value, col3: value}` = 1 + 3 + 3 + 3-1 + 1
= 1`{` + 3 `col1: value ,` + 3 `col2: value` + 2 `col3: value` + 1`}`

File: cpp/src/io/json/write_json.cu (hunk 2 above)
Comment (karthikeyann) [reply cpp/src/io/json/write_json.cu]: I noticed this early. The entire file could use nosync update. Created PR #17445

**Review by vuule — APPROVED** [review]

**Pull request merged**